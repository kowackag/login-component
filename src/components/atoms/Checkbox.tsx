import { FC, forwardRef, DetailedHTMLProps, InputHTMLAttributes } from "react";

export type CheckboxTypes = {
	id: string;
	name: string;
	label: string;
	// type?: HTMLInputElement;
	className?: string;
} & DetailedHTMLProps<InputHTMLAttributes<HTMLInputElement>, HTMLInputElement>;

export const Checkbox: FC<CheckboxTypes> = forwardRef<HTMLInputElement, CheckboxTypes>(
	({ id, name, label, className, ...props }, ref) => {
		return (
			<div className="justify-beetwen mb-4 flex align-middle">
				<input
					id={id}
					ref={ref}
					name={name}
					type="checkbox"
					className={`peer fixed h-5 w-5 opacity-0 ${className ?? ""}`}
					{...props}
				/>
				<p className="mr-3 inline-block h-5 w-5 rounded-sm border border-neutral-300 p-0 leading-5 text-white peer-checked:border-none peer-checked:bg-bright-rose ">
					<svg width="16px" height="16px">
						<use href="/icons.svg#check"></use>
					</svg>
				</p>
				<label className="font-primaryMedium leading-5" htmlFor={name}>
					{label}
				</label>
			</div>
		);
	},
);
