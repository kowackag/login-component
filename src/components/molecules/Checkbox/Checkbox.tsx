import { ErrorMessage } from "@/components/atoms/Error/Error";
import { ChangeEventHandler, FocusEventHandler, Ref } from "react";
import { Path } from "react-hook-form";

export interface IFormValues {
	email: string;
	password: number;
	toSave: boolean;
}

export type CheckboxFieldTypes = {
	name: Path<IFormValues>;
	label: string;
	onChange: ChangeEventHandler;
	onBlur: FocusEventHandler<HTMLInputElement>;
	inputRef: Ref<HTMLInputElement>;
	error: string | undefined;
};

export const Checkbox = ({
	name,
	label,
	onChange,
	onBlur,
	inputRef,
	error,
}: CheckboxFieldTypes) => {
	return (
		<div className="justify-beetwen mb-4 flex align-middle">
			<input
				id={name}
				name={name}
				type="checkbox"
				onChange={onChange}
				onBlur={onBlur}
				ref={inputRef}
				className="peer fixed h-5 w-5 opacity-0"
			/>
			<p className="mr-3 inline-block h-5 w-5 rounded-sm border border-neutral-300 p-0 leading-5 text-white peer-checked:border-none peer-checked:bg-bright-rose ">
				<svg width="16px" height="16px">
					<use href="/icons.svg#check"></use>
				</svg>
			</p>
			<label htmlFor={name} className="font-primaryMedium leading-5">
				{label}
			</label>
			{error ? <ErrorMessage message={error} /> : null}
		</div>
	);
};
