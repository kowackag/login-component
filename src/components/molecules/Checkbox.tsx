import { ErrorMessage } from "@/components/atoms/Error";
import { CheckboxFieldTypes } from "@/utils/shared.types";

export const Checkbox = ({
	name,
	label,
	onChange,
	onBlur,
	inputRef,
	error,
}: CheckboxFieldTypes) => {
	return (
		<div className="justify-beetwen mb-4 flex align-middle">
			<input
				id={name}
				name={name}
				type="checkbox"
				onChange={onChange}
				onBlur={onBlur}
				ref={inputRef}
				className="peer fixed h-5 w-5 opacity-0"
			/>
			<p className="mr-3 inline-block h-5 w-5 rounded-sm border border-neutral-300 p-0 leading-5 text-white peer-checked:border-none peer-checked:bg-bright-rose ">
				<svg width="16px" height="16px">
					<use href="/icons.svg#check"></use>
				</svg>
			</p>
			<label htmlFor={name} className="font-primaryMedium leading-5">
				{label}
			</label>
			{error ? <ErrorMessage message={error} /> : null}
		</div>
	);
};
