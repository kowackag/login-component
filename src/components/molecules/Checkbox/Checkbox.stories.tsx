import type { Meta, StoryObj } from "@storybook/react";

import { Checkbox } from "./Checkbox";

const meta: Meta<typeof Checkbox> = {
	component: Checkbox,
	decorators: [
		(Story) => (
			<div style={{ width: "360px" }}>
				<Story />
			</div>
		),
	],
};

export default meta;

type Story = StoryObj<typeof Checkbox>;

export const Default: Story = {
	args: {
		label: "Save?",
		error: "it is required",
	},
};
