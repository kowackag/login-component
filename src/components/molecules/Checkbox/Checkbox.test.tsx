import { describe, expect, it, vi } from "vitest";
import { fireEvent, render, screen } from "@testing-library/react";
import { Checkbox } from "./Checkbox";

describe("Checkbox", () => {
	it("component exist", () => {
		const mockFn = vi.fn();
		const { container } = render(
			<Checkbox
				name="toSave"
				label="Save?"
				error=""
				onChange={mockFn}
				onBlur={mockFn}
				inputRef={mockFn}
			/>,
		);

		const checkbox = container.querySelector("#toSave") as HTMLInputElement;
		expect(checkbox).toBeInTheDocument();
	});
	it("has appropriate type", () => {
		const mockFn = vi.fn();
		const { container } = render(
			<Checkbox
				name="toSave"
				label="Save?"
				error=""
				onChange={mockFn}
				onBlur={mockFn}
				inputRef={mockFn}
			/>,
		);

		const checkbox = container.querySelector("#toSave") as HTMLInputElement;
		expect(checkbox.type).toBe("checkbox");
	});
	it("changes if is selected", () => {
		const mockFn = vi.fn();
		const { container } = render(
			<Checkbox
				name="toSave"
				label="Save?"
				error=""
				onChange={mockFn}
				onBlur={mockFn}
				inputRef={mockFn}
			/>,
		);
		const checkbox = [...container.querySelectorAll("#toSave")][0] as HTMLInputElement;

		fireEvent.click(checkbox);
		expect(mockFn).toBeCalled();
		expect(checkbox.checked).toBe(true);
	});
	it("checkbox error is shown with correct text", () => {
		const mockFn = vi.fn();
		render(
			<Checkbox
				name="toSave"
				label="Save?"
				error="some error"
				onChange={mockFn}
				onBlur={mockFn}
				inputRef={mockFn}
			/>,
		);
		const errorEl = screen.getByTestId("error");
		expect(errorEl).toBeVisible();
		expect(errorEl).toHaveTextContent("some error");
	});
	it("input error is not rendered", () => {
		const mockFn = vi.fn();
		const { container } = render(
			<Checkbox
				name="toSave"
				label="Save?"
				error=""
				onChange={mockFn}
				onBlur={mockFn}
				inputRef={mockFn}
			/>,
		);
		const error = container.querySelector("[data-testid='error']") as HTMLInputElement;

		expect(error).toBe(null);
	});
	it("renders correctly", () => {
		const mockFn = vi.fn();
		const wrapper = render(
			<Checkbox
				name="toSave"
				label="Save?"
				error=""
				onChange={mockFn}
				onBlur={mockFn}
				inputRef={mockFn}
			/>,
		);
		expect(wrapper.container).toMatchSnapshot();
	});
});
