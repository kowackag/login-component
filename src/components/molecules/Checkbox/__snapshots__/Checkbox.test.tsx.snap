// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Checkbox > renders correctly 1`] = `
<div>
  <div
    class="justify-beetwen relative mb-4 flex align-middle"
  >
    <input
      class="peer fixed h-5 w-5 opacity-0"
      data-testid="checkbox-toSave"
      id="toSave"
      name="toSave"
      type="checkbox"
    />
    <p
      class="mr-3 inline-block h-5 w-5 rounded-sm border border-neutral-300 p-0 leading-5 text-white peer-checked:border-none peer-checked:bg-bright-rose "
    >
      <svg
        height="16px"
        width="16px"
      >
        <use
          href="/icons.svg#check"
        />
      </svg>
    </p>
    <label
      class="font-primaryMedium leading-5"
      for="toSave"
    >
      Save?
    </label>
  </div>
</div>
`;
